### PR TITLE
refactor(template-compiler): Remove parent AST references

### PR DESCRIPTION
**Pull request by jmsjtu**
## Details
This PR covers the following two tasks from the [Github issue](https://github.com/salesforce/lwc/issues/2432):

- Remove parent field from IRNodes: The future AST nodes won't have parent back pointers (because it introduces circular dependencies). We can take a head start by changing the compiler code to keep track of the parent chain instead of relying on the information stored on the AST node.

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

If yes, please describe the impact and migration path for existing applications.

## The PR fulfills these requirements:
* Have tests for the proposed changes been added? ✅ 
* Have you followed [these instructions](../CONTRIBUTING.md#-commit-message-conventions) to clearly describe the issue being fixed or feature enhanced? ✅ 

## GUS work item
W-9698817

### CODE REVIEW

File: packages/@lwc/template-compiler/src/parser/index.ts
Comment (pmdartus) [inline packages/@lwc/template-compiler/src/parser/index.ts:1028]: I would expect `getRoot` to be:

```suggestion
        return parentStack[0];
```

It is currently needed when `getRoot` is invoked on the root node itself. We can keep this for now, but it's something we should refactor after landing the new AST format.

File: packages/@lwc/template-compiler/src/shared/ir.ts
Comment (pmdartus) [inline packages/@lwc/template-compiler/src/shared/ir.ts:94]: Instead of using a while loop, you can use a simple for:

```js
for (let i = parentStack.length - 1; i >= 0; i--) {
  current = parentStack[i];
  // [...]
} 
```

File: packages/@lwc/template-compiler/src/parser/expression.ts
Comment (pmdartus) [inline packages/@lwc/template-compiler/src/parser/expression.ts:157]: I don't understand why this is needed. Accessing an array out of bound value should return `undefined`.

**Review by pmdartus — CHANGES_REQUESTED** [review]

**Review by ekashida — APPROVED** [review]
lgtm, let's merge after addressing PM's comments

File: packages/@lwc/template-compiler/src/parser/expression.ts
Comment (jmsjtu) [reply packages/@lwc/template-compiler/src/parser/expression.ts:157]: Great feedback!  I didn't know that about arrays in javascript, TIL!

File: packages/@lwc/template-compiler/src/shared/ir.ts
Comment (jmsjtu) [reply packages/@lwc/template-compiler/src/shared/ir.ts]: I updated isInIteration from parser/index.ts and isComponentProp to use the for loop.

File: packages/@lwc/template-compiler/src/parser/expression.ts
Comment (pmdartus) [inline packages/@lwc/template-compiler/src/parser/expression.ts:127]: I am surprised that we have to do `String.prototype.toLowerCase()`. The tag should already be in lowercase.
I ran the test locally and it is not needed.

File: packages/@lwc/template-compiler/src/parser/expression.ts
Comment (pmdartus) [inline packages/@lwc/template-compiler/src/parser/expression.ts:143]: Ditto.

File: packages/@lwc/template-compiler/src/codegen/index.ts
Comment (pmdartus) [inline packages/@lwc/template-compiler/src/codegen/index.ts:66]: Instead of wrapping all the `transformChildren` invocation with a stack push and a pop, I would rather recommend pushing the element to the stack directly inside `transformChildren`. This would make the code more self-contained.

```ts
function transformChildren(parent: IRNode, children: IRNode[]): t.Expression {
  // ...
  parentStack.push(parent);
  while ((current = childrenIterator.next()) && !current.done) {
    // ...
  }
  parentStack.pop();
  // ...
}
```

**Review by pmdartus — APPROVED** [review]
Added a couple of extra feedback. Once addressed LGTM!

**Pull request merged**
